packages/hooks/src/use-mutation-observer/use-mutation-observer.ts

```diff
@@ -0,0 +1,100 @@
+// NOTE: This React hook is essentially the same as the [@react-hook/intersection-observer](https://www.npmjs.com/package/@react-hook/intersection-observer)
+// except for the usage of `MutationObserver` instead of `IntersectionObserver`.
+import useLatest from '@react-hook/latest';
+import rafSchd from 'raf-schd';
+import { useLayoutEffect } from 'react';
+
+type TUseMutationObserverCallback = (
+  mutationsList: MutationRecord[],
+  observer: MutationObserver
+) => void;
+
+let _mutationObserver: ReturnType<typeof createMutationObserver>;
+
+function createMutationObserver() {
+  const callbacks = new Map<Node, TUseMutationObserverCallback[]>();
+  const observer = new MutationObserver(
+    rafSchd((mutationsList, observer) => {
+      const mutationsByTarget = mutationsList.reduce<
```

Comment:
We group the mutations by target and deliver the complete list to the consumer callback instead of calling it once per mutation.
This is closer to how the [MutationObserver API works](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/MutationObserver#parameters).